nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord.java

```diff
@@ -0,0 +1,253 @@
+                writer.flush();
+            }
+            records.add(baos.toByteArray());
+        }
+        return new InFlightMessageMonitor(records);
+    }
+
+    protected void sendAsync(Producer<byte[]> producer, ProcessSession session, FlowFile flowFile, InFlightMessageMonitor monitor) {
+        if (monitor == null || monitor.getRecords().isEmpty())
+           return;
+
+        for (byte[] record: monitor.getRecords() ) {
+           try {
+              publisherService.submit(new Callable<Object>() {
+                @Override
+                public Object call() throws Exception {
+                  try {
+                     return producer.sendAsync(record).handle((msgId, ex) -> {
+                         if (msgId != null) {
+                            monitor.getSuccessCounter().incrementAndGet();
+                            return msgId;
+                         } else {
+                            monitor.getFailureCounter().incrementAndGet();
+                            monitor.getFailures().add(record);
+                            return null;
+                         }
+                     }).get();
+
+                   } catch (final Throwable t) {
+                      // This traps any exceptions thrown while calling the producer.sendAsync() method.
+                      monitor.getFailureCounter().incrementAndGet();
+                      monitor.getFailures().add(record);
+                      return null;
+                   } finally {
+                      monitor.getLatch().countDown();
+                   }
+               }
+             });
+          } catch (final RejectedExecutionException ex) {
+            // This can happen if the processor is being Unscheduled.
+          }
+       }
+    }
+
+    private void handleAsync(InFlightMessageMonitor monitor, ProcessSession session, FlowFile flowFile, String topic) {
+       try {
+           boolean useOriginalForFailures = false;
+           monitor.getLatch().await();
+
+           if (monitor.getSuccessCounter().intValue() > 0) {
+               flowFile = session.putAttribute(flowFile, MSG_COUNT, monitor.getSuccessCounter().get() + "");
+               flowFile = session.putAttribute(flowFile, TOPIC_NAME, topic);
+               session.adjustCounter("Messages Sent", monitor.getSuccessCounter().get(), true);
+               session.getProvenanceReporter().send(flowFile, "Sent " + monitor.getSuccessCounter().get() + " records to " + topic );
+               session.transfer(flowFile, REL_SUCCESS);
+           } else {
+              // Route the original FlowFile to failure, otherwise we need to create a new FlowFile for the Failure relationship
+              useOriginalForFailures = true;
+           }
+
+           if (monitor.getFailureCounter().intValue() > 0) {
+              // Create a second flow file for failures.
+              FlowFile failureFlowFile = useOriginalForFailures ? flowFile : session.create();
```

Comment:
We should always transfer the original FlowFile to failure if there is a failure - never a new FlowFile.